cmd/mailbox-controller/controller.go

```diff
@@ -253,63 +263,119 @@ func (ctl *mbCtl) sync(ctx context.Context, refany any) bool {
-		return true
+
+	// The script must be idempotent.
+	shellScriptName := "kubestellar-kube-bind"
+
+	resourcesToBind := []string{"syncerconfigs", "edgesyncconfigs"}
+	for idx, resource := range resourcesToBind {
+		logger.V(2).Info("Ensuring binding", "script", shellScriptName, "resource", resource)
+
+		// suffix helps isolate this controller's multiple workers to prevent concurrent access of a single kubeconfig file.
+		// This is a compromise due to the situation that we are using kcp which modifies kubeconfig file when changing workspaces.
+		freshSuffix := atomic.AddUint64(&suffix, 1)
+
+		makeCopyOfKubeConfig := fmt.Sprintf("cp $KUBECONFIG $KUBECONFIG.copy%d", freshSuffix)
+		removeCopyWhenExits := fmt.Sprintf("trap 'rm $KUBECONFIG.copy%d' EXIT", freshSuffix)
+		invokeScript := strings.Join([]string{
+			fmt.Sprintf("KUBECONFIG=$KUBECONFIG.copy%d", freshSuffix),
+			shellScriptName,
+			workspace.Name,
+			resource,
+		}, " ")
+		if idx == 0 {
+			invokeScript = invokeScript + " --start-konnector true"
+		}
+		cmdLine := strings.Join([]string{
+			makeCopyOfKubeConfig,
+			removeCopyWhenExits,
+			invokeScript,
+		}, "; ")
+		logger.V(2).Info("About to exec", "cmdLine", cmdLine)
+		cmd := exec.Command("/bin/sh", "-c", cmdLine)
+		stdout, err := cmd.StdoutPipe()
+		if err != nil {
+			logger.Error(err, "Failed to extract stdout pipe for exec of script", "script", shellScriptName)
+			return true
+		}
+		stderr, err := cmd.StderrPipe()
+		if err != nil {
+			logger.Error(err, "Failed to extract stderr pipe for exec of script", "script", shellScriptName)
+			return true
+		}
+		err = cmd.Start()
+		if err != nil {
+			logger.Error(err, "Unable to start executing the script", "script", shellScriptName)
+			return true
+		}
+		outBuf := bufio.NewReader(stdout)
+		errBuf := bufio.NewReader(stderr)
+		for {
+			line, _, err := outBuf.ReadLine()
+			if err != nil {
+				if err == io.EOF {
+					logger.V(4).Info("End of stdout")
+					break
+				} else {
+					logger.Error(err, "Unable to read stdout")
+					return true
+				}
+			}
+			logger.V(2).Info("Stdout from exec", "resource", resource, "line", line)
+		}
+		for {
+			line, _, err := errBuf.ReadLine()
+			if err != nil {
+				if err == io.EOF {
+					logger.V(4).Info("End of stderr")
+					break
+				} else {
+					logger.Error(err, "Unable to read stderr")
+					return true
+				}
+			}
+			logger.V(2).Info("Stderr from exec", "resource", resource, "line", line)
+		}
+		if err = cmd.Wait(); err != nil {
+			logger.Error(err, "Unable to bind", "workspace", workspace.Name, "resrouce", resource)
+			return true
+		}
 	}
-	logger.V(2).Info("Created APIBinding", "resourceVersion", binding2.ResourceVersion)
 	return false
 }
 
-func mbwsNameOfSynctarget(st *edgev2alpha1.SyncTarget) string {
-	cluster := logicalcluster.From(st)
-	return cluster.String() + wsNameSep + string(st.UID)
+func (ctl *mbCtl) mbwsNameOfSynctarget(st *edgev2alpha1.SyncTarget) string {
```

Comment:
There are two semantically distinct failure cases in this method. One is not going to change when retried later, the other probably is. The caller cares about this difference.

Reply:
done